4337/contracts/test/SafeSignerLaunchpad.sol

```diff
@@ -0,0 +1,266 @@
+ * @dev The is intended to be set as a Safe proxy's implementation for ERC-4337 user operation that deploys the account.
+ */
+contract SafeSignerLaunchpad is IAccount, SafeStorage {
+    bytes32 private constant DOMAIN_SEPARATOR_TYPEHASH = keccak256("EIP712Domain(uint256 chainId,address verifyingContract)");
+
+    // keccak256("SafeSignerLaunchpad.initHash") - 1
+    uint256 private constant INIT_HASH_SLOT = 0x1d2f0b9dbb6ed3f829c9614e6c5d2ea2285238801394dc57e8500e0e306d8f80;
+
+    /**
+     * @notice The keccak256 hash of the EIP-712 SafeInit struct, representing the structure of a ERC-4337 compatible deferred Safe initialization.
+     *  {address} singleton - The singleton to evolve into during the setup.
+     *  {address} signerFactory - The unique signer factory to use for creating an owner.
+     *  {bytes} signerData - The signer data to use the owner.
+     *  {address} setupTo - The contract to delegatecall during setup.
+     *  {bytes} setupData - The calldata for the setup delegatecall.
+     *  {address} fallbackHandler - The fallback handler to initialize the Safe with.
+     */
+    bytes32 private constant SAFE_INIT_TYPEHASH =
+        keccak256(
+            "SafeInit(address singleton,address signerFactory,bytes signerData,address setupTo,bytes setupData,address fallbackHandler)"
+        );
+
+    /**
+     * @notice The keccak256 hash of the EIP-712 SafeInitOp struct, representing the user operation to execute alongside initialization.
+     *  {bytes32} userOpHash - The user operation hash being executed.
+     *  {uint48} validAfter - A timestamp representing from when the user operation is valid.
+     *  {uint48} validUntil - A timestamp representing until when the user operation is valid, or 0 to indicated "forever".
+     *  {address} entryPoint - The address of the entry point that will execute the user operation.
+     */
+    bytes32 private constant SAFE_INIT_OP_TYPEHASH =
+        keccak256("SafeInitOp(bytes32 userOpHash,uint48 validAfter,uint48 validUntil,address entryPoint)");
+
+    address private immutable SELF;
+    address public immutable SUPPORTED_ENTRYPOINT;
+
+    constructor(address entryPoint) {
+        require(entryPoint != address(0), "Invalid entry point");
+
+        SELF = address(this);
+        SUPPORTED_ENTRYPOINT = entryPoint;
+    }
+
+    modifier onlyProxy() {
+        require(singleton == SELF, "Not called from proxy");
+        _;
+    }
+
+    modifier onlySupportedEntryPoint() {
+        require(msg.sender == SUPPORTED_ENTRYPOINT, "Unsupported entry point");
+        _;
+    }
+
+    receive() external payable {}
+
+    function setup(bytes32 initHash, address to, bytes calldata preInit) external onlyProxy {
+        _setInitHash(initHash);
+        if (to != address(0)) {
+            (bool success, ) = to.delegatecall(preInit);
+            require(success, "Pre-initialization failed");
+        }
+    }
```

Comment:
Just to confirm I understand things correctly: this shall be called through the factory during the deployment?

Reply:
nit: i know that the arity is different but it's too similar to the safe's core `setup` function. I'd name it `preValidationSetup` to be more descriptive

Reply:
Good point - I will rename it.

Reply:
Renamed in cf404bfb60f270ee9b1bd50c5a49a5f4aed2efad